core/drivers/crypto/stm32/stm32_cryp.c

```diff
@@ -0,0 +1,1304 @@
+#define _CRYP_CR_GCM_CCMPH_HEADER	1U
+#define _CRYP_CR_GCM_CCMPH_PAYLOAD	2U
+#define _CRYP_CR_GCM_CCMPH_FINAL	3U
+#define _CRYP_CR_CRYPEN			BIT(15)
+#define _CRYP_CR_FFLUSH			BIT(14)
+#define _CRYP_CR_KEYSIZE_MSK		GENMASK_32(9, 8)
+#define _CRYP_CR_KEYSIZE_OFF		8U
+#define _CRYP_CR_KSIZE_128		0U
+#define _CRYP_CR_KSIZE_192		1U
+#define _CRYP_CR_KSIZE_256		2U
+#define _CRYP_CR_DATATYPE_MSK		GENMASK_32(7, 6)
+#define _CRYP_CR_DATATYPE_OFF		6U
+#define _CRYP_CR_DATATYPE_NONE		0U
+#define _CRYP_CR_DATATYPE_HALF_WORD	1U
+#define _CRYP_CR_DATATYPE_BYTE		2U
+#define _CRYP_CR_DATATYPE_BIT		3U
+#define _CRYP_CR_ALGOMODE_MSK		(BIT(19) | GENMASK_32(5, 3))
+#define _CRYP_CR_ALGOMODE_OFF		3U
+#define _CRYP_CR_ALGOMODE_TDES_ECB	0x0U
+#define _CRYP_CR_ALGOMODE_TDES_CBC	0x1U
+#define _CRYP_CR_ALGOMODE_DES_ECB	0x2U
+#define _CRYP_CR_ALGOMODE_DES_CBC	0x3U
+#define _CRYP_CR_ALGOMODE_AES_ECB	0x4U
+#define _CRYP_CR_ALGOMODE_AES_CBC	0x5U
+#define _CRYP_CR_ALGOMODE_AES_CTR	0x6U
+#define _CRYP_CR_ALGOMODE_AES		0x7U
+#define _CRYP_CR_ALGOMODE_AES_GCM	BIT(16)
+#define _CRYP_CR_ALGOMODE_AES_CCM	(BIT(16) | BIT(0))
+#define _CRYP_CR_ALGODIR		BIT(2)
+#define _CRYP_CR_ALGODIR_ENC		0U
+#define _CRYP_CR_ALGODIR_DEC		BIT(2)
+
+/* CRYP status register fields */
+#define _CRYP_SR_BUSY			BIT(4)
+#define _CRYP_SR_OFFU			BIT(3)
+#define _CRYP_SR_OFNE			BIT(2)
+#define _CRYP_SR_IFNF			BIT(1)
+#define _CRYP_SR_IFEM			BIT(0)
+
+/* CRYP DMA control register fields */
+#define _CRYP_DMACR_DOEN		BIT(1)
+#define _CRYP_DMACR_DIEN		BIT(0)
+
+/* CRYP interrupt fields */
+#define _CRYP_I_OUT			BIT(1)
+#define _CRYP_I_IN			BIT(0)
+
+/* CRYP hardware configuration register fields */
+#define _CRYP_HWCFGR_CFG1_MSK		GENMASK_32(3, 0)
+#define _CRYP_HWCFGR_CFG1_OFF		0U
+#define _CRYP_HWCFGR_CFG2_MSK		GENMASK_32(7, 4)
+#define _CRYP_HWCFGR_CFG2_OFF		4U
+#define _CRYP_HWCFGR_CFG3_MSK		GENMASK_32(11, 8)
+#define _CRYP_HWCFGR_CFG3_OFF		8U
+#define _CRYP_HWCFGR_CFG4_MSK		GENMASK_32(15, 12)
+#define _CRYP_HWCFGR_CFG4_OFF		12U
+
+/* CRYP HW version register */
+#define _CRYP_VERR_MSK			GENMASK_32(7, 0)
+#define _CRYP_VERR_OFF			0U
+
+/*
+ * Macro to manage bit manipulation when we work on local variable
+ * before writing only once to the real register.
+ */
+#define CLRBITS(v, bits)		((v) &= ~(bits))
+#define SETBITS(v, bits)		((v) |= (bits))
+static inline void clrsetbits(uint32_t *v, uint32_t mask, uint32_t bits)
+{
+	*v = (*v & ~mask) | bits;
+}
+
+#define IS_ALGOMODE(cr, mod) \
+	(((cr) & _CRYP_CR_ALGOMODE_MSK) == (_CRYP_CR_ALGOMODE_##mod << \
+					  _CRYP_CR_ALGOMODE_OFF))
+
+#define SET_ALGOMODE(mod, cr) \
+	clrsetbits(&(cr), _CRYP_CR_ALGOMODE_MSK, (_CRYP_CR_ALGOMODE_##mod << \
+						  _CRYP_CR_ALGOMODE_OFF))
+
+#define GET_ALGOMODE(cr) \
+	(((cr) & _CRYP_CR_ALGOMODE_MSK) >> _CRYP_CR_ALGOMODE_OFF)
+
+#define TOBE32(x)			TEE_U32_BSWAP(x)
+#define FROMBE32(x)			TEE_U32_BSWAP(x)
+#define BE_INDEX(i)			({ typeof(i) _i = (i);      \
+					(_i / 4) * 4 + 3 - _i % 4; })
```

Comment:
for consistency and readability, i think a local helper function would be better:
```c
static size_t be_index(uint32_t index)
{ 
 	return (index & ~0x3) + 3 - (index & 0x3);
}